### ISSUE TITLE
12-nlp-url-spam.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/crerov-intro-machine-learning#9]
@crerov 

### **Cumplimiento de Requisitos**

| Paso | Requisito | Estado | Detalle |
|------|-----------|--------|---------|
| 1 | Cargar url_spam.csv | ✅ | Correcto |
| 2 | Preprocesar URLs | ✅ | Tokenización, lematización, stopwords, minlen=4 |
| 2 | Split 80/20 | ⚠️ | Sin `stratify=y` |
| 2 | Vectorización TF-IDF | ⚠️ | Fit ANTES del split (data leakage) |
| 3 | SVM base | ✅ | `kernel="linear"` correcto |
| 3 | Análisis resultados | ❌ | **SOLO accuracy, SIN classification_report** |
| 4 | Optimización | ⚠️ | GridSearchCV, pero parámetros confusos |
| 5 | Guardar modelo | ⚠️ | **FALTA VECTORIZADOR ❌** |

---

#### **1. CRÍTICO - Data leakage en vectorización**

Tu código:
```python
# ← TF-IDF ajustado CON TODO EL DATASET
vectorizer = TfidfVectorizer(max_features=5000, max_df=0.8, min_df=5)
X = vectorizer.fit_transform(url_strings).toarray()
y = df["is_spam"]

# ← Después divides
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```

**Problema:** Vectorizador "ve" test set al calcular IDF.

**Código correcto:**
```python
# 1. Split PRIMERO
X_train_urls, X_test_urls, y_train, y_test = train_test_split(
    url_strings, y,
    test_size=0.2,
    random_state=42,
    stratify=y  # ← Añade esto también
)

# 2. Fit TF-IDF SOLO en train
vectorizer = TfidfVectorizer(max_features=5000, max_df=0.8, min_df=5)
X_train = vectorizer.fit_transform(X_train_urls).toarray()
X_test = vectorizer.transform(X_test_urls).toarray()  # Solo transform
```

---

#### **2. CRÍTICO - NO guardas el VECTORIZADOR**

Tu código:
```python
with open("/workspaces/crerov-intro-machine-learning/models/NLP.pkl", 'wb') as f:
    pickle.dump(svm_model, f)
```

**Guarda solo el modelo, FALTA el vectorizador.**

**En producción, ¿cómo vectorizas nuevas URLs?**

```python
# Esto FALLARÍA en producción:
model = pickle.load("NLP.pkl")
new_url = ["https://example.com"]
# ¿Con qué vectorizador transformo new_url? NO EXISTE
X_new = vectorizer.transform([new_url])  #  Error: vectorizer no está guardado
prediction = model.predict(X_new)
```

**Debería ser:**
```python
# Guardar AMBOS
with open("svm_model.pkl", 'wb') as f:
    pickle.dump(svm_model, f)
with open("tfidf_vectorizer.pkl", 'wb') as f:
    pickle.dump(vectorizer, f)

# En producción:
model = pickle.load("svm_model.pkl")
vectorizer = pickle.load("tfidf_vectorizer.pkl")
X_new = vectorizer.transform([preprocess_and_clean(new_url)])
prediction = model.predict(X_new)
```

---

#### **3. IMPORTANTE - Sin stratify en split**

```python
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# ← FALTA stratify=y
```

Con desbalance ~89.7% no-spam vs 10.3% spam:
- Podría quedar train: 550 spam, 1840 no-spam
- Test: 146 spam, 463 no-spam (diferente proporción)

**Correcto:**
```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2, 
    random_state=42,
    stratify=y  # ← Mantiene proporción
)
```

---

#### **4. MENOR - Sin visualizaciones**

No hay:
- Confusion matrices
- Gráficos de comparación base vs optimizado
- Curvas de rendimiento

Otros estudiantes (Anais, Alessandra, Ana, Natalia) incluyen estos.

---

### **Análisis de la "optimización"**

Tu conclusión dice:
```
"El modelo optimizado no mejora, guardamos el original"
```

**Esta decisión es correcta**, pero:

**Hipótesis:** ¿Por qué la optimización empeora?
1. **Data leakage** - El vectorizador "vio" el test set, inflando accuracy base
2. **Overfitting base** - Modelo base sobreajustado, optimización lo regulariza
3. **Desbalance** - Accuracy es métrica pobre para desbalance
4. **GridSearchCV limitado** - 72 combinaciones quizá no suficientes

**Sin métricas por clase, es imposible saber.**
